### DISCUSSION TITLE
How to get taghelpers to use route area instead of module name?

### DISCUSSION THREAD

**Discussion by inzi** [OrchardCMS/OrchardCore#11126] [Q&A]
In my manifest, I have a module called "My.Admin.SMS". 

The purpose of this is to allow a user who doesn't have access to admin panel in Orchard to manage some SMS settings if they have the correct role, along with some other modules. 

So I have the regular modules rendering correctly. However, I have a page with a different layout in the theme.

To reach these these tools like this: myorchardsite.com/vmadmin

In my theme's layout, if the request path starts with /vmadmin I use a different page layout.

In that page, I have a link to smsadmin (myorchardsite.com/vmadmin/smsadmin)

In my controller, I have route(s) set: 
[Route("vmadmin/smsadmin")]

I have a table with an edit link per row.

<a asp-action="editprovider" asp-route-id="@model.id">Edit</a>

On first request, an asp-action for an anchor tag renders the URL as http://myorchardsite.com/My.Admin.SMS/sms/editprovider/55 (where Sms is the controller name [SmsController] and 55 is the ID).

If I modify the url to an endpoint http://myorchardsite.com/vmadmin/smsadmin/editprovider/55, the area changes and the taghelper's asp-action renders as http://myorchardsite.com/vmadmin/smsadmin
 from the page, the taghelper renders the URL as http://myorchardsite.com/vmadmin/

In examining the request object, on the initial request, the area is the module name, the action is Index, and the controller is Sms.

None of the taghelper generated links work (404 given) until the area is correct. 

How do I get the module to return the "friendly" area so tag-helpers work as expected in OrchardCore?  Using taghelpers automatically generate antiforgery information.

Thanks for any guidance!

**Comment by inzi** [comment]
I think I've almost found my answer.. 

1 - Add attribute base route to controller. [Route("vmadmin/smsadmin"]
2 - Add the MapAreaControllerRoute with name route and pattern with actions and optional id
3 - Use asp-route="smsadmin", then do asp-route-action and asp-route-id to build the url.

However - this only works in the module for smsadmin.

I've tried adding this MapAreaControllerRoute to other modules, but the tag helpers don't work on any page that's not rendered by the sms controller.

Any ideas?

I've tried the theme and other modules that link to it.

**Reply by Piedone** [reply]
I'm quite lost on what you're trying to achieve and what you have. Can you clear this up by providing code snippets of what you have in your modules, what output it produces, and what you want to achieve instead?

**Comment by ns8482e** [comment]
Add `asp-area="My.Admin.SMS"` in your tag helper to correctly generate
